File: sdk-java/src/main/java/com/crosscert/passkey/sdk/internal/TraceIdPropagationInterceptor.java
package com.crosscert.passkey.sdk.internal;

import com.crosscert.passkey.sdk.PasskeyClientConfig;
import org.springframework.http.HttpRequest;
import org.springframework.http.client.ClientHttpRequestExecution;
import org.springframework.http.client.ClientHttpRequestInterceptor;
import org.springframework.http.client.ClientHttpResponse;
import java.io.IOException;

/**
 * 호출자의 traceId(기본=MDC)를 X-Trace-Id 로 전파. RedactingRequestInterceptor 보다
 * 먼저 등록되어야 redaction DEBUG 로그 시점에 헤더가 존재한다.
 */
public class TraceIdPropagationInterceptor implements ClientHttpRequestInterceptor {

    private static final String TRACE_HEADER = "X-Trace-Id";

    private final PasskeyClientConfig config;

    public TraceIdPropagationInterceptor(PasskeyClientConfig config) {
        this.config = config;
    }

    @Override
    public ClientHttpResponse intercept(HttpRequest request, byte[] body,
                                        ClientHttpRequestExecution exec) throws IOException {
        String traceId = config.traceIdProvider().get();
        if (traceId != null && !traceId.isBlank()) {
            request.getHeaders().set(TRACE_HEADER, traceId);
        }
        return exec.execute(request, body);
    }
}

File: sdk-java/src/main/kotlin/com/crosscert/passkey/sdk/internal/TraceIdPropagationInterceptor.kt
package com.crosscert.passkey.sdk.internal

import com.crosscert.passkey.sdk.PasskeyClientConfig
import org.springframework.http.HttpRequest
import org.springframework.http.client.ClientHttpRequestExecution
import org.springframework.http.client.ClientHttpRequestInterceptor
import org.springframework.http.client.ClientHttpResponse
import java.io.IOException

/**
 * Propagates the caller's `traceId` (read via
 * [PasskeyClientConfig.traceIdProvider], default = MDC) onto the
 * outgoing SDK request as `X-Trace-Id`. The receiving passkey-app's
 * `TraceIdFilter` re-uses the inbound header so the two servers
 * share a single trace id for one logical user action.
 *
 * This interceptor must be registered BEFORE [RedactingRequestInterceptor]
 * so the header is present when redaction logging runs (the redactor
 * inspects request headers for its DEBUG output).
 */
class TraceIdPropagationInterceptor(
    private val config: PasskeyClientConfig,
) : ClientHttpRequestInterceptor {

    @Throws(IOException::class)
    override fun intercept(
        request: HttpRequest,
        body: ByteArray,
        exec: ClientHttpRequestExecution,
    ): ClientHttpResponse {
        // traceIdProvider 는 플랫폼 타입이라 런타임에 null 을 반환할 수 있으므로 nullable 로 받는다.
        val traceId: String? = config.traceIdProvider.get()
        if (traceId != null && !traceId.isBlank()) {
            request.headers.set(TRACE_HEADER, traceId)
        }
        return exec.execute(request, body)
    }

    companion object {
        private const val TRACE_HEADER = "X-Trace-Id"
    }
}
